synapse/rest/client/v1/login.py

```diff
@@ -339,6 +340,27 @@ async def _complete_login(
                 )
             user_id = canonical_uid
 
+        if login_submission.get("org.matrix.msc2697.restore_device"):
+            (
+                device_id,
+                dehydrated_device,
+            ) = await self.device_handler.get_dehydrated_device(user_id)
+            if dehydrated_device:
+                token = await self.device_handler.get_dehydration_token(
+                    user_id, device_id, login_submission
+                )
+                result = {
+                    "user_id": user_id,
+                    "home_server": self.hs.hostname,
+                    "device_data": dehydrated_device,
+                    "device_id": device_id,
+                    "dehydration_token": token,
+                }
+
+                # FIXME: call callback?
```

Comment:
The [documentation about what this callback might be](https://github.com/matrix-org/synapse/blob/master/docs/password_auth_providers.md#optional-methods) is less than ideal. It comes from password auth modules and might be `None` or a `Callable` that is "called with the result from the `/login` call (including `access_token`, `device_id`, etc.)"

I do not know of any auth providers which use this frankly, but I think calling it like below is the reasonable thing to do. Is there a particular reason you think it should not be called?

Reply:
We talked a bit more about this and the reason we can't just call it here is that we don't yet have the access token (the client hasn't acknowledged that they've been able to use the dehydrated device).

It seems like we would ideally want to call this at the end of the call to `/restore_device`, but we no longer have the callback at that point. Will need to think about what the best option is here.